dateutil/parser.py

```diff
@@ -742,318 +763,215 @@ def _parse(self, timestr, dayfirst=None, yearfirst=None, fuzzy=False,
-
-                            i += 1
+                                    raise InvalidDatetimeError(timestr)
 
-                            if i < len_l and l[i] == sep:
+                            if i+3 < len_l and l[i+3] == sep:
                                 # We have three members
-                                i += 1
-                                value = info.month(l[i])
+                                value = info.month(l[i+4])
 
                                 if value is not None:
-                                    ymd.append(value)
-                                    assert mstridx == -1
-                                    mstridx = len(ymd)-1
+                                    ymd.append(value, 'M')
                                 else:
-                                    ymd.append(l[i])
-
-                                i += 1
-                    elif i >= len_l or info.jump(l[i]):
-                        if i+1 < len_l and info.ampm(l[i+1]) is not None:
-                            # 12 am
-                            res.hour = int(value)
+                                    ymd.append(l[i+4])
+                                i += 2
 
-                            if res.hour < 12 and info.ampm(l[i+1]) == 1:
-                                res.hour += 12
-                            elif res.hour == 12 and info.ampm(l[i+1]) == 0:
-                                res.hour = 0
+                            i += 1
+                        i += 1
 
+                    elif i+1 >= len_l or info.jump(l[i+1]):
+                        if i+2 < len_l and info.ampm(l[i+2]) is not None:
+                            # 12 am
+                            hour = int(value)
+                            res.hour = _adjust_ampm(hour, info.ampm(l[i+2]))
                             i += 1
                         else:
                             # Year, month or day
                             ymd.append(value)
                         i += 1
-                    elif info.ampm(l[i]) is not None:
 
+                    elif info.ampm(l[i+1]) is not None:
                         # 12am
-                        res.hour = int(value)
-
-                        if res.hour < 12 and info.ampm(l[i]) == 1:
-                            res.hour += 12
-                        elif res.hour == 12 and info.ampm(l[i]) == 0:
-                            res.hour = 0
+                        hour = int(value)
+                        res.hour = _adjust_ampm(hour, info.ampm(l[i+1]))
                         i += 1
 
                     elif not fuzzy:
-                        return None, None
-                    else:
-                        i += 1
-                    continue
+                        raise InvalidDatetimeError(timestr)
+
 
                 # Check weekday
-                value = info.weekday(l[i])
-                if value is not None:
+                elif info.weekday(l[i]) is not None:
```

Comment:
> Was the huge if statement being pointlessly broken up just so that info.weekday(l[i]) didn't need to be called twice?

AFAICT, yes.  I was pretty sure that the big `if` blocks were mutually exclusive, but the number of `continue` statements, `return (None, None)`, and exceptions made for enough code paths that I wanted to `(n+1)`-check.  I chose to take the performance hit of a few extra dict lookups in order to make this obvious and have only one non-raising path through the loop.